### ISSUE TITLE
Error handling and response decoding

### ISSUE THREAD

**Issue by sagikazarmark** [toin0u/concise#4]
There are two things we can handle:
- Http errors, bad responses, which is probably handled by the adapter and returned as a unified error
- Validate the returned data. If the returned data contains error message, throw a Provider exception, otherwise invalid data.

I would say let Concise fail when some sort of adapter/http issue occurs and only handle invalid/faulty data. Or we can handle all errors (exceptions) and say we can have Unknown, Provider, and Invalid data exceptions.

@toin0u What do you think? This is rather related to API usage which you have more experience in than me.

**Comment by toin0u** [comment]
I think it's important to catch exceptions and re-throws them in the Concise context. It will provide better error messages. I'll work on this :)

**Comment by sagikazarmark** [comment]
Yeah, I agree that the exceptions should be handled in there. But it is still a question, what sort of errors and how we handle them.

Looking forward to your solution ;)

**Comment by sagikazarmark** [comment]
@toin0u any progress?

**Comment by sagikazarmark** [comment]
The way how responses are decoded should be part of this. For example if a JSON response is returned, and it is decoded to an object, it should be checked if the appropriate properties exist. (However I prefer JSON to be decoded to array)

**Comment by toin0u** [comment]
I really like to do this but no time :-/ and I was sick too.. It's in my todo list :)

I agree with your comment but why do you prefer JSON decoded to array ?

**Comment by sagikazarmark** [comment]
`json_decode('<json>', true)`

**Comment by toin0u** [comment]
Yes :) but why do you prefer to `json_decode('<json>')` ?

**Comment by sagikazarmark** [comment]
That's just a replacement for a JSON data object. Or do you mean why I
prefer JSON for REST API?

**Comment by toin0u** [comment]
Sorry my bad. I was wondering why do you prefer to convert the JSON into an associative array by doing `json_decode('<json>', true)` instead of using the `stdClass` got with `json_decode('<json>')` :)

**Comment by sagikazarmark** [comment]
Array is just a bunch of data with an optional structure. Objects are a bit more, a higher level. So I tend to keep these kind of data in arrays.

**Comment by toin0u** [comment]
I completely forgot to work on that.. Sorry for that @sagikazarmark x_X

**Comment by sagikazarmark** [comment]
No problem. Does that mean you are not intended to work on it anymore?

**Comment by toin0u** [comment]
No no :) I just need to postpone it later on..
